controllers/provider-aws/pkg/controlplane/internal/helper_test.go

```diff
@@ -0,0 +1,227 @@
+			decoder.EXPECT().Decode(cp.Spec.InfrastructureProviderStatus.Raw, nil, &apisaws.InfrastructureStatus{}).DoAndReturn(decode)
+
+			// Create helper
+			h := NewHelper()
+
+			// Call ConfigChartValuesFunc and check the results
+			values, err := h.ConfigChartValuesFunc(decoder, cp, cluster)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(values).To(Equal(expectedValues))
+		})
+	})
+
+	Describe("#ControlPlaneChartValuesFunc", func() {
+		It("should return correct control plane chart values", func() {
+			var (
+				checksums = map[string]string{
+					cloudControllerManagerDeploymentName: "1",
+					cloudControllerManagerServerName:     "2",
+					common.CloudProviderSecretName:       "3",
+					common.CloudProviderConfigName:       "4",
+				}
+
+				expectedValues = map[string]interface{}{
+					"cloudProvider":     "aws",
+					"clusterName":       namespace,
+					"kubernetesVersion": "1.13.4",
+					"podNetwork":        cidr,
+					"replicas":          1,
+					"podAnnotations": map[string]interface{}{
+						"checksum/secret-cloud-controller-manager":        "1",
+						"checksum/secret-cloud-controller-manager-server": "2",
+						"checksum/secret-cloudprovider":                   "3",
+						"checksum/configmap-cloud-provider-config":        "4",
+					},
+					"configureRoutes": false,
+					"environment": []map[string]interface{}{
+						{
+							"name": "AWS_ACCESS_KEY_ID",
+							"valueFrom": map[string]interface{}{
+								"secretKeyRef": map[string]interface{}{
+									"key":  aws.AccessKeyID,
+									"name": common.CloudProviderSecretName,
+								},
+							},
+						},
+						{
+							"name": "AWS_SECRET_ACCESS_KEY",
+							"valueFrom": map[string]interface{}{
+								"secretKeyRef": map[string]interface{}{
+									"key":  aws.SecretAccessKey,
+									"name": common.CloudProviderSecretName,
+								},
+							},
+						},
+					},
+					"resources": map[string]interface{}{
+						"limits": map[string]interface{}{
+							"cpu":    "500m",
+							"memory": "512Mi",
+						},
+					},
+					"featureGates": map[string]bool{
+						"CustomResourceValidation": true,
+					},
+				}
+			)
+
+			// Create mock decoder
+			decoder := mockruntime.NewMockDecoder(ctrl)
+			decoder.EXPECT().Decode(cp.Spec.ProviderConfig.Raw, nil, &apisaws.ControlPlaneConfig{}).DoAndReturn(decode)
+
+			// Create helper
+			h := NewHelper()
+
+			// Call ControlPlaneChartValuesFunc and check the results
+			values, err := h.ControlPlaneChartValuesFunc(decoder, cp, cluster, checksums)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(values).To(Equal(expectedValues))
+		})
+	})
+
+})
+
+func encode(obj runtime.Object) []byte {
+	data, _ := json.Marshal(obj)
+	return data
+}
+
+func decode(data []byte, _ *schema.GroupVersionKind, obj runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+	var err error
+	switch x := obj.(type) {
+	case *apisaws.ControlPlaneConfig:
```

Comment:
Why this switch? You're anyways always doing the same.

Reply:
The switch is a way to convert `obj` into a concrete type, so that `Unmarshal` works correctly. I know it looks strange, but it's the only way to write it - you need different cases for `x` to change its type, even though the 2 cases look the same. You can try it out if you have doubts.

Reply:
As said, this switch definitely has no value. You could just directly call `json.Unmarshal`.